go/vt/topo/metadata.go

```diff
@@ -0,0 +1,106 @@
+/*
+Copyright 2017 Google Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package topo
+
+import (
+	"context"
+	"fmt"
+	"path"
+	"regexp"
+	"strings"
+
+	"vitess.io/vitess/go/event"
+	"vitess.io/vitess/go/vt/topo/events"
+)
+
+// UpsertMetadata sets the key/value in the metadata if it doesn't exist, otherwise it updates the content
+func (ts *Server) UpsertMetadata(ctx context.Context, key string, val string) error {
+	keyPath := path.Join(MetadataPath, key)
+	for {
+		_, _, err := ts.globalCell.Get(ctx, keyPath)
+		switch {
+		case err == nil:
+			return update(ctx, ts, keyPath, val)
+		case IsErrType(err, NoNode):
+			return create(ctx, ts, keyPath, val)
+		default:
+			return err
+		}
+	}
+}
+
+// GetMetadata retrieves all metadata value that matches the given key regular expression. If empty all values are returned.
+func (ts *Server) GetMetadata(ctx context.Context, keyFilter string) (map[string]string, error) {
+	keys, err := ts.globalCell.ListDir(ctx, MetadataPath, false)
+	if err != nil {
+		return nil, err
+	}
+
+	keyPattern := strings.Replace(regexp.QuoteMeta(keyFilter), "%", ".*", -1)
+	keyPattern = fmt.Sprintf("^%s$", keyPattern)
+	re := regexp.MustCompile(keyPattern)
```

Comment:
I'd add a little comment at the end like `// Can never fail.`

Reply:
Done